core/src/main/java/org/apache/iceberg/SnapshotManager.java

```diff
@@ -82,6 +92,7 @@ public ManageSnapshots cherrypick(long snapshotId) {
       // link the snapshot about to be published on commit with the picked snapshot
       set(SnapshotSummary.SOURCE_SNAPSHOT_ID_PROP, String.valueOf(targetSnapshotId));
     } else {
+      //TODO: snapshotOperation maybe unset here.
```

Comment:
Yes, because an existing snapshot is used, `operation` should not be called.

Reply:
Can we  reliably set `requiresSnapshotCreation` when operation is being configured, e.g a fast-forward can be converted to a cherry-pick only during apply which can result in snapshot creation, no?

Reply:
Hm. Good point. Maybe it is cleaner to set this in apply then.

Reply:
Oops, this should have been a reply to another comment. Fixing it.

Reply:
This is also affect the Operation?. If a cherry-pick  can be fast forward-able in apply or a fast-forward can be converted to a cherry-pick . How does this affect the `operation()` api?

Reply:
Operation is only called internally, when creating the snapshot summary. When the commit is a fast-forward, it isn't called because no new snapshot is created. That's why we didn't worry about it and made sure that if `operation` was called it would fail.

Here's an alternative that might be simpler: use the parent's `snapshotId` method and check whether there is a snapshot with that ID. If there is, then this created a new snapshot by cherry-picking. If there isn't, then it was a fast-forward to the `targetSnapshotId`. By overriding the `updateEvent` method, you can detect everything you need to without worrying about making these return "real" values.

Reply:
I think this sounds like a much better plan. Let me try it.